File: sceneview/src/main/java/com/google/ar/sceneform/rendering/ExternalTexture.java
package com.google.ar.sceneform.rendering;

import android.graphics.SurfaceTexture;
import android.view.Surface;

import androidx.annotation.Nullable;

import com.google.android.filament.Engine;
import com.google.android.filament.Stream;
import com.google.android.filament.Texture;
import io.github.sceneview.collision.Preconditions;

/**
 * Creates an Android {@link SurfaceTexture} and {@link Surface} that can be displayed by Sceneform.
 * Useful for displaying video, or anything else that can be drawn to a {@link SurfaceTexture}.
 *
 * <p>The getFilamentEngine OpenGL ES texture is automatically created by Sceneform. Also, {@link
 * SurfaceTexture#updateTexImage()} is automatically called and should not be called manually.
 *
 * <p>Call {@link Material#setExternalTexture(String, ExternalTexture)} to use an ExternalTexture.
 * The material parameter MUST be of type 'samplerExternal'.
 */
public class ExternalTexture {
    private static final String TAG = ExternalTexture.class.getSimpleName();

    @Nullable
    private final SurfaceTexture surfaceTexture;
    @Nullable
    private final Surface surface;

    @Nullable
    private Texture filamentTexture;
    @Nullable
    private Stream filamentStream;

    private Engine engine;

    /**
     * Creates an ExternalTexture with a new Android {@link SurfaceTexture} and {@link Surface}.
     */
    @SuppressWarnings("initialization")
    public ExternalTexture(Engine engine) {
        this.engine = engine;
        SurfaceTexture surfaceTexture = new SurfaceTexture(0);
        surfaceTexture.detachFromGLContext();
        this.surfaceTexture = surfaceTexture;

        // Create the Android surface.
        surface = new Surface(surfaceTexture);

        // Create the filament stream.
        //TODO : We actually have an issue with stream being destroyed from elsewhere maybe material
        // Fix it whith kotlining here
        Stream stream = new Stream.Builder().stream(surfaceTexture).build(engine);

        //TODO : We actually have an issue with stream being destroyed from elsewhere maybe material
        // Fix it when kotlining here
        initialize(stream);
    }

    @SuppressWarnings("initialization")
    private void initialize(Stream filamentStream) {
        if (filamentTexture != null) {
            throw new AssertionError("Stream was initialized twice");
        }

        // Create the filament stream.
        this.filamentStream = filamentStream;

        // Create the filament texture.
        filamentTexture = new Texture.Builder()
                .sampler(Texture.Sampler.SAMPLER_EXTERNAL)
                .format(Texture.InternalFormat.RGB8)
                .build(engine);

        filamentTexture.setExternalStream(engine, filamentStream);
    }

    /**
     * Gets the surface texture created for this ExternalTexture.
     */
    public SurfaceTexture getSurfaceTexture() {
        return Preconditions.checkNotNull(surfaceTexture);
    }

    /**
     * Gets the surface created for this ExternalTexture that draws to {@link #getSurfaceTexture()}
     */
    public Surface getSurface() {
        return Preconditions.checkNotNull(surface);
    }

    public Texture getFilamentTexture() {
        return Preconditions.checkNotNull(filamentTexture);
    }

    public Stream getFilamentStream() {
        return Preconditions.checkNotNull(filamentStream);
    }

    public void destroy() {
        if (filamentTexture != null) {
            engine.destroyTexture(filamentTexture);
        }
        filamentTexture = null;

        if (filamentStream != null) {
            engine.destroyStream(filamentStream);
        }
        filamentStream = null;
    }
}

File: sceneview/src/main/java/com/google/ar/sceneform/rendering/ExternalTexture.kt
package com.google.ar.sceneform.rendering

import android.graphics.SurfaceTexture
import android.view.Surface
import com.google.android.filament.Engine
import com.google.android.filament.Stream
import com.google.android.filament.Texture
import io.github.sceneview.collision.Preconditions

/**
 * Creates an Android [SurfaceTexture] and [Surface] that can be displayed by Sceneform.
 * Useful for displaying video, or anything else that can be drawn to a [SurfaceTexture].
 *
 * The getFilamentEngine OpenGL ES texture is automatically created by Sceneform. Also,
 * [SurfaceTexture.updateTexImage] is automatically called and should not be called manually.
 *
 * Call [Material.setExternalTexture] to use an ExternalTexture.
 * The material parameter MUST be of type 'samplerExternal'.
 */
class ExternalTexture(private val engine: Engine) {

    private val surfaceTexture: SurfaceTexture?
    private val surface: Surface?

    private var filamentTexture: Texture? = null
    private var filamentStream: Stream? = null

    init {
        val st = SurfaceTexture(0)
        st.detachFromGLContext()
        surfaceTexture = st

        // Create the Android surface.
        surface = Surface(st)

        // Create the filament stream.
        //TODO : We actually have an issue with stream being destroyed from elsewhere maybe material
        // Fix it with kotlining here
        val stream = Stream.Builder().stream(st).build(engine)

        //TODO : We actually have an issue with stream being destroyed from elsewhere maybe material
        // Fix it when kotlining here
        initialize(stream)
    }

    private fun initialize(filamentStream: Stream) {
        if (filamentTexture != null) {
            throw AssertionError("Stream was initialized twice")
        }

        // Create the filament stream.
        this.filamentStream = filamentStream

        // Create the filament texture.
        filamentTexture = Texture.Builder()
            .sampler(Texture.Sampler.SAMPLER_EXTERNAL)
            .format(Texture.InternalFormat.RGB8)
            .build(engine)

        filamentTexture!!.setExternalStream(engine, filamentStream)
    }

    /** Gets the surface texture created for this ExternalTexture. */
    fun getSurfaceTexture(): SurfaceTexture = Preconditions.checkNotNull(surfaceTexture)

    /** Gets the surface created for this ExternalTexture that draws to [getSurfaceTexture] */
    fun getSurface(): Surface = Preconditions.checkNotNull(surface)

    fun getFilamentTexture(): Texture = Preconditions.checkNotNull(filamentTexture)

    fun getFilamentStream(): Stream = Preconditions.checkNotNull(filamentStream)

    fun destroy() {
        filamentTexture?.let { engine.destroyTexture(it) }
        filamentTexture = null

        filamentStream?.let { engine.destroyStream(it) }
        filamentStream = null
    }

    companion object {
        private val TAG = ExternalTexture::class.java.simpleName
    }
}
